### PR TITLE
generate keepSourceAttribution from RawKeep

### PR DESCRIPTION
**Pull request by yingjieMiao**
@andrewconner @stkem 

note I add a column to bookmark table. Cache version is not touched.

### CODE REVIEW

File: kifi-backend/modules/common/conf/evolutions/shoebox/304.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# SHOEBOX
+
+# --- !Ups
+
+alter table bookmark
+    add column source_attribution_id bigint(20) NULL;
```
Comment (stkem) [inline kifi-backend/modules/common/conf/evolutions/shoebox/304.sql:6]: foreign key?

**Comment by stkem** [comment]
lgtm. we need to be careful with this table alter, do to the size. Better do it in the evening.

File: kifi-backend/modules/common/conf/evolutions/shoebox/304.sql (hunk 1 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/conf/evolutions/shoebox/304.sql:6]: good catch

**Comment by yingjieMiao** [comment]
I agree with you on the size concern. On the other hand, the table has some manual written sql queries plus  db-row-to-object mapping, i.e., things like (`keptAt = r.<<[DateTime]`). If this kind of stuff goes wrong in the night...

Anyway, I will do some localhost test to be double sure.

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala [hunk 2]
```diff
@@ -30,7 +30,8 @@ case class Keep(
     kifiInstallation: Option[ExternalId[KifiInstallation]] = None,
     seq: SequenceNumber[Keep] = SequenceNumber.ZERO,
     libraryId: Option[Id[Library]],
-    keptAt: DateTime = currentDateTime) extends ModelWithExternalId[Keep] with ModelWithState[Keep] with ModelWithSeqNumber[Keep] {
+    keptAt: DateTime = currentDateTime,
+    sourceAttributionId: Option[Id[KeepSourceAttribution]] = None) extends ModelWithExternalId[Keep] with ModelWithState[Keep] with ModelWithSeqNumber[Keep] {
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala:34]: Thanks for adding it to the end of the class. Makes it easier.

File: kifi-backend/modules/shoebox/app/com/keepit/model/RawKeep.scala [hunk 3]
```diff
@@ -27,6 +27,17 @@ case class RawKeep(
   def withUpdateTime(now: DateTime) = this.copy(updatedAt = now)
 }
 
+object RawKeep {
+  def extractKeepSourceAttribtuion(keep: RawKeep): Option[KeepSourceAttribution] = {
+    keep.source match {
+      case KeepSource.twitterFileImport | KeepSource.twitterSync =>
+        val attrOpt = keep.originalJson.flatMap(js => TwitterAttribution.fromRawTweetJson(js))
+        attrOpt.map { attr => KeepSourceAttribution(attribution = attr) }
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/RawKeep.scala:37]: Mind logging if it fails? No alert needed, but we'd like to have some mechanism of knowing if it's not working.

**Comment by andrewconner** [comment]
Thanks for adding the tests. Looks good to me.

**Comment by yingjieMiao** [comment]
added foreign key and logging

**Pull request merged**